open-notes/maturity-model.md

```diff
@@ -0,0 +1,394 @@
+
+### Learning days
+
+Once low hanging fruit has been tackled it is important for people to have more freedom to learn and experiment to get gains. Having a day dedicated (say every second week) to learning will foster innovation that can often fold back into the product. 
+
+### Synthetics
+
+By having controlled tests running that simulate users in production it is possible to know with greater certainty what the user is experiencing. It is important that these are measured from the outside but can be tied to telemetry.
+
+### Cyclomatic complexity ceiling
+
+A lot of code metrics and recommendations can be debated but cyclomatic complexity is a fairly straightforward measure of the complexity. Measuring this and putting a ceiling on what is acceptable before refactoring efforts need to be focused in that area.
+
+### Quantitative team health
+
+Find ways to measure how the team is doing and visualize these. Then identify when the trend is going poorly and take action to turn things around. The idea here is 
+
+- Code/test metrics
+- Aspect rating
+
+### Embrace FP principles
+
+Functions and immutable data. Distinguish between calculations (pure functions) and actions (impure functions). 
+Keep the level of abstraction consistent at different layers (stratified design).
+
+### PR to production
+
+Each PR to the main branch ends up in production when merged with zero manual intervention (and within a few minutes of the merge). This implies:
+
+- trust in tests
+- impeccable monitoring and alerting
+- fast recovery/rollback
+
+### Dashboards IV
+
+The final dashboard includes non-system elements like:
+
+- team metrics
+- code metrics
+
+![Metrics framework](images/maturity-model-2021-07-28-19-40-55.png)
+
+ Then bring the dashboards together across system business performance, infrastructure performance, delivery performance, and team health.
+
+ ### Load testing
+
+ Automated test runs can be scheduled to test how the systems perform under heavy load.
+
+ ### Immutable Architecture
+
+- Versioned API
+- Idempotent API
+- Non-destructive data mutations only
+- Event feeds
+
+### Domain Driven design
+
+- Ubiquitous language is shared and found in the codebase (Value types as a starting point)
+- Strong understanding from development of processes and workflows (use EventStorming)
+- Shared model with business (aggregates enforce business invariance and repositories return valid aggregates only). Model business process over structure.
+- Use cases as Application Service (preferred)
+
+For the rest use patterns as needed but keep it simple.
+
+### Property based testing
+
+Gain for confidence by testing the full range of a domain. Discover the characteristics of your system.
+
+### Chaos engineering
+
+Build resilient systems by exercising the resilience often. Automate partial outages of systems to dependent systems are forced to evolve to be stronger.
```

Comment:
Might be interesting the mechanisms you might then need to think about. As Joao mentioned, graceful degradation, which involves retrying, maybe a fallback option, and ultimately (partial) failure.